### PR TITLE
[EEG Browser] Signal Visualization, Events, Electrode map and EEG split files support

### PR DESCRIPTION
**Pull request by laemtl**
This PR adds dynamic visualization of EEG based on a previous project started by Armin Taheri (https://github.com/aces/react-series-data-viewer). 
These changes and include cleanup of older code in the Electrophysiology Browser and new features like viewing the list of Events, and an Electrode Map dynamically plotted from the BIDS-compliant coordinates file. 

The visualization component uses svg (Visx|D3) and introduces a new stack of dependencies:

### Ramda (https://ramdajs.com)
A practical functional library for JavaScript programmers.

### Redux (https://redux.js.org)
A Predictable State Container for JS Apps

### Visx (https://airbnb.io/visx)
A collection of expressive, low-level visualization primitives for React.

### RxJS (https://rxjs-dev.firebaseapp.com/guide/overview)
RxJS is a library for composing asynchronous and event-based programs by using observable sequences. 
It provides one core type, the Observable, satellite types (Observer, Schedulers, Subjects) and operators to allow handling asynchronous events as collections.

### TypeScript (https://www.typescriptlang.org)
A static type checker for javascript.

## Preview 
Screenshot_2021-04-09 LORIS Demonstration Database - Electrophysiology Browser - View Session(6)

## To test 
- install protoc: 
    - `sudo apt-get install -y protobuf-compiler` (Ubuntu > 16.10) 
    - `brew install protobuf` (Mac)
    -  https://github.com/protocolbuffers/protobuf/blob/master/src/README.md
- run npm install
- The MRI pipeline must be run on raisinbread to get the signal chunks
- To view all the functionalities, an event and electrode file is needed in the BIDS package.

## To discuss
This PR contains the [protobuf file](https://github.com/aces/Loris/blob/3b2703e9a970b7ef2377d4048380a0fbcd6df99d/modules/electrophysiology_browser/jsx/react-series-data-viewer/protocol-buffers/chunk.proto) used to generate the frontend (js) and backend (python, used in Loris MRI) chunk message encoder/parser. I'm not sure this file should be stored here and what would be a good compilation pipeline.  

## To fix
The electrode map needs to support a coordinate system to perform appropriate rotations.

### CODE REVIEW

**Comment by laemtl** [comment]
Blocked by #7345

**Comment by laemtl** [comment]
Annotation functionality is blocked by #7345

**Comment by christinerogers** [comment]
to update @laemtl :  Test plan and Module Readme.  dd

**Review by cmadjar — COMMENTED** [review]
@laemtl I cannot see the div with the signal when I checkout your branch. Below is what I see.

Screen Shot 2021-03-24 at 11 26 18 AM

In the console, I see the following errors:
```
Failed to load resource: http://localhost:8080/mri/jiv/get_file.php?file=bids_imports/Face13_BIDSVersion_1.1.0_chunks/sub-OTT170_ses-V1_task-faceO_eeg.chunks/index.json the server responded with a status of 404 (Not Found)
```
(I did set the `imagePath` config to the correct path so it should have found it as I can `ls` it on my computer and the permissions appear to be fine as well)

```
Error: Request failed with status code 404 EEGLabSeriesProvider.js:66
    at createError (createError.js:16)
    at settle (settle.js:18)
    at XMLHttpRequest.handleLoad (xhr.js:59)
```

Let me know if I can provide more information to debug this. A quick call with share screen is also possible :).

**Comment by laemtl** [comment]
@cmadjar I implemented a new file system to replace `/mri/jiv/get_file.php`. That should solve your issue.

**Review by cmadjar — CHANGES_REQUESTED** [review]
@laemtl This looks so good! This is supra reactive. Very impressive!!

A few comments:
- on the signal graph, the 0 is cut in its display. Is it possible to show it so that it is not cut in half
- when I click on the arrows of the "Showing" box, I get something funny. See screenshot
Screen Shot 2021-04-06 at 10 16 32 AM
- I see a lot of white space in the File Download div and in the channel label column of the channel and epoch viewer. Is there a way to decrease that so that there is more room for the interesting stuff (signal and electrode location)
  - other possibility, in the File download div, have the summary table as well to take some of the empty space and the details div could contain the details table only. So the File download could be relabelled File summary information and download? 
- How can I modify the timeline displayed? I see a timeline range reset button but I cannot figure out how to modify the time scale or even navigate through the time scale.
- I tried clicking on the electrode map. 
  - Is it supposed to update the channel and epoch viewer with the signal of that electrode?
  - could the line in the table showing the electrode name be highlighted when clicking on an electrode on the brain and vice versa?

**Comment by christinerogers** [comment]
fyi @cmadjar there's still a few tweaks on the cosmetic side for this tool 
e.g. moving a few buttons/messages around, like the "too many channels to display" message, optimizing the whitespace where possible.  
(@laemtl let me know if you want to go over again what's outstanding)

**Comment by laemtl** [comment]
`How can I modify the timeline displayed? I see a timeline range reset button but I cannot figure out how to modify the time scale or even navigate through the time scale.`

Screenshot from 2021-04-06 14-07-02

Both ends of the timeline bar (in pale blue) are draggable to limit the time range.

**Comment by laemtl** [comment]
```
when I click on the arrows of the "Showing" box, I get something funny. See screenshot
Screen Shot 2021-04-06 at 10 16 32 AM
```

I may be wrong, but it seems specific to your browser. I'm unable to reproduce on Firefox and Chrome.

**Comment by laemtl** [comment]
`on the signal graph, the 0 is cut in its display. Is it possible to show it so that it is not cut in half` - fixed

**Comment by cmadjar** [comment]
@laemtl is this ready for another round of testing?

**Comment by cmadjar** [comment]
@laemtl your branch is now conflicting with main. It literally just conflicted under my eyes ;)

**Comment by laemtl** [comment]
I will rebase shortly 😊

**Comment by laemtl** [comment]
@cmadjar Ready for second review.

**Comment by cmadjar** [comment]
@laemtl I get the following error when I run `make dev`

```
ERROR in ./modules/electrophysiology_browser/jsx/react-series-data-viewer/src/series/components/SeriesRenderer.js 159:55
    Module parse failed: Unexpected token (159:55)
    File was processed with these loaders:
     * ./node_modules/babel-loader/lib/index.js
    You may need an additional loader to handle the result of these loaders.
    |       stroke: "black"
    |     }), filteredChannels.map(function (channel, i) {
    >       var seriesRange = channelMetadata[channel.index]?.seriesRange;
    |       if (!seriesRange) return null;
    |       return React.createElement(Axis, {
```

**Comment by laemtl** [comment]
Hum weird, we might have a different setup. It's fine I can rewrite this section.

File: modules/electrophysiology_browser/jsx/react-series-data-viewer/src/series/components/EventManager.js
Comment (maltheism) [inline modules/electrophysiology_browser/jsx/react-series-data-viewer/src/series/components/EventManager.js:32]: @laemtl I think the <> isn't needed since one div encompasses all the other elements.

File: modules/electrophysiology_browser/jsx/react-series-data-viewer/src/series/components/SeriesRenderer.js
Comment (maltheism) [inline modules/electrophysiology_browser/jsx/react-series-data-viewer/src/series/components/SeriesRenderer.js:336]: Hi @laemtl, I think the <> isn't needed here either.

File: modules/electrophysiology_browser/jsx/react-series-data-viewer/src/series/components/SeriesRenderer.js
Comment (maltheism) [inline modules/electrophysiology_browser/jsx/react-series-data-viewer/src/series/components/SeriesRenderer.js:557]: @laemtl should this commented out code be kept in here?

Comment (laemtl) [reply modules/electrophysiology_browser/jsx/react-series-data-viewer/src/series/components/SeriesRenderer.js:557]: It's a temporary hack, since the feature is not ready yet, but yeah, it's cleaner to remove it for now.

**Comment by maltheism** [comment]
@laemtl browser console warnings when using react development:
```
Warning: Failed prop type: Invalid prop `value` supplied to `SelectElement`.
SelectElement@http://localhost/js/components/Form.js:642:81
Filter@http://localhost/electrophysiology_browser/js/electrophysiologyBrowserIndex.js:3628:58
div
div
div
Panel@http://localhost/electrophysiology_browser/js/electrophysiologyBrowserIndex.js:2029:81
FilterableDataTable@http://localhost/electrophysiology_browser/js/electrophysiologyBrowserIndex.js:1659:81
ElectrophysiologyBrowserIndex@http://localhost/electrophysiology_browser/js/electrophysiologyBrowserIndex.js:153:81

Warning: Failed prop type: Invalid prop `value` of type `boolean` supplied to `TextboxElement`, expected `string`.
TextboxElement@http://localhost/js/components/Form.js:1270:81
Filter@http://localhost/electrophysiology_browser/js/electrophysiologyBrowserIndex.js:3628:58
div
div
div
Panel@http://localhost/electrophysiology_browser/js/electrophysiologyBrowserIndex.js:2029:81
FilterableDataTable@http://localhost/electrophysiology_browser/js/electrophysiologyBrowserIndex.js:1659:81
ElectrophysiologyBrowserIndex@http://localhost/electrophysiology_browser/js/electrophysiologyBrowserIndex.js:153:81
```

**Comment by maltheism** [comment]
Hi @laemtl, when clicking the "Download Table as CSV" button and viewing the csv will not display the links correctly in the csv.

**Review by maltheism — CHANGES_REQUESTED** [review]
Hi @laemtl, see comments.

File: modules/electrophysiology_browser/jsx/components/electrophysiology_session_summary.js
Comment (HenriRabalais) [inline modules/electrophysiology_browser/jsx/components/electrophysiology_session_summary.js:1]: This file appears to be tabbed 4 spaces instead of 2.

File: modules/electrophysiology_browser/jsx/react-series-data-viewer/src/series/components/ResponsiveViewer.js
Comment (HenriRabalais) [inline modules/electrophysiology_browser/jsx/react-series-data-viewer/src/series/components/ResponsiveViewer.js:28]: Is this a way of declaring prop types/default props? If so, I like it!

**Review by HenriRabalais — COMMENTED** [review]
Took a look at all the javascript! I've never worked with any of the dependencies that are introduced, so most of the syntax I'm not familiar with and so for the most part it's a bit above my head. Code looks clean though!

The one thing I would suggest is to take the components that could be generalized loris-wide (e.g. SummaryPanel, DownloadPanel, FilePanel, Sidebar, SidebarContent, etc.) and find a way to integrate them in the main `loris/jsx` folder. I think this will help standardize components across loris and avoid duplication of components in the future.

File: modules/electrophysiology_browser/jsx/react-series-data-viewer/src/series/components/ResponsiveViewer.js
Comment (laemtl) [reply modules/electrophysiology_browser/jsx/react-series-data-viewer/src/series/components/ResponsiveViewer.js:28]: Yes, but it's a syntax specific to flow. Typescript is more popular nowadays.

**Review by cmadjar — COMMENTED** [review]
@laemtl This is looking so good!! A few things came to mind when I looked at the GUI. I think several of them will be features for later. Maybe it would be good to convert some of the comments below into issues to not forget. I am happy with how the GUI looks already. This is amazing!!

A few things I noticed along the way:
- **timeline section**: change the mouse pointer to become a hand so users understand that they can play with that timeline. Also have a tooltip on the timeline showing to explain to users how to play with the timeline section of the graph
- **events section**: When the mouse is over the event box on the right, highlight the event on the graph. Right now, it does that only when on top of the eye icon but the eye icon also make the event disappearing from the graph when clicked on. Maybe add a tool tip as well on top of the eye with "Show event" or "Do not show event" depending on the status of the button
- **graph section**: when on a window small enough to see the events, we can see the name of the event displayed at the button. 👍  Would it be possible to also highlight the event in the events panel at the same time?
- **head section**: 
  - I wonder whether the highlighting of the electrode when we mouse over  the name of the electrode could not be more obvious? Maybe a different colour that appears more obvious to the eye?
  - I do not see any highlighting when on the 3D view. I guess this is for later?
  - when we mouse over an electrode on the head, it might be good to highlight the row on the table next showing electrode names to easily see the name of the electrode.

**Review by cmadjar — APPROVED** [review]
The changes mentioned in my previous review will be done on separate PRs and moved into an issue since the vis is already looking really good :).

**Comment by laemtl** [comment]
> @laemtl browser console warnings when using react development and visiting `http://localhost/electrophysiology_browser/`:
> 
> ```
> Warning: Failed prop type: Invalid prop `value` supplied to `SelectElement`.
> SelectElement@http://localhost/js/components/Form.js:642:81
> Filter@http://localhost/electrophysiology_browser/js/electrophysiologyBrowserIndex.js:3628:58
> div
> div
> div
> Panel@http://localhost/electrophysiology_browser/js/electrophysiologyBrowserIndex.js:2029:81
> FilterableDataTable@http://localhost/electrophysiology_browser/js/electrophysiologyBrowserIndex.js:1659:81
> ElectrophysiologyBrowserIndex@http://localhost/electrophysiology_browser/js/electrophysiologyBrowserIndex.js:153:81
> 
> Warning: Failed prop type: Invalid prop `value` of type `boolean` supplied to `TextboxElement`, expected `string`.
> TextboxElement@http://localhost/js/components/Form.js:1270:81
> Filter@http://localhost/electrophysiology_browser/js/electrophysiologyBrowserIndex.js:3628:58
> div
> div
> div
> Panel@http://localhost/electrophysiology_browser/js/electrophysiologyBrowserIndex.js:2029:81
> FilterableDataTable@http://localhost/electrophysiology_browser/js/electrophysiologyBrowserIndex.js:1659:81
> ElectrophysiologyBrowserIndex@http://localhost/electrophysiology_browser/js/electrophysiologyBrowserIndex.js:153:81
> ```

This problem needs to be fixed for all module. Maybe a good fix for a separate PR :)

**Comment by laemtl** [comment]
> Hi @laemtl, when clicking the "Download Table as CSV" button and viewing the csv will not display the links correctly in the csv.

This is a good comment for PR #7345

**Comment by christinerogers** [comment]
> The changes mentioned in my previous review will be done on separate PRs and moved into an issue since the vis is already looking really good :).

Love your review points @cmadjar - we will definitely do a round 2 to finesse some of the controls, especially the time interval selection.  
I'm collecting additional user feedback from other corners as well these days, to add for round 2, including things I"m noticing late like the button colour to show events is a stark dark blue...

**Review by maltheism — APPROVED** [review]

**Comment by laemtl** [comment]
@driusan This PR is ready for you!

**Comment by laemtl** [comment]
Blocked by #7345

**Comment by cmadjar** [comment]
@laemtl this is failing the test. Could you fix it?

**Comment by driusan** [comment]
@laemtl ping

**Comment by laemtl** [comment]
@cmadjar, @driusan Fixed.

**Comment by laemtl** [comment]
@maltheism Do you have time to review?

**Comment by maltheism** [comment]
@laemtl one quick question I thought of while I'm now testing this PR. You have protobuf support for macOS and ubuntu. I'm just wondering if centos that LORIS supports would also have something for protobuf?

**Comment by maltheism** [comment]
@laemtl when visiting http://localhost/electrophysiology_browser/ I get the following in browser console:
Warning: Failed prop type: Invalid prop `value` supplied to `SelectElement`.

Warning: Failed prop type: Invalid prop `value` of type `boolean` supplied to `TextboxElement`, expected `string`.

**Comment by laemtl** [comment]
@maltheism Yes, the thirs link 

> @laemtl when visiting http://localhost/electrophysiology_browser/ I get the following in browser console:
> Warning: Failed prop type: Invalid prop `value` supplied to `SelectElement`.
> 
> Warning: Failed prop type: Invalid prop `value` of type `boolean` supplied to `TextboxElement`, expected `string`.

Good point but unrelated to my changes in this PR. I suggest to open an issue to address those errors in all modules.

**Comment by laemtl** [comment]
> @laemtl one quick question I thought of while I'm now testing this PR. You have protobuf support for macOS and ubuntu. I'm just wondering if centos that LORIS supports would also have something for protobuf?

Yes, for CentOS users can use the third link: 

install protoc:

    sudo apt-get install -y protobuf-compiler (Ubuntu > 16.10)
    brew install protobuf (Mac)
    https://github.com/protocolbuffers/protobuf/blob/master/src/README.md

**Comment by maltheism** [comment]
Hi @laemtl, thanks for answers and you're right I didn't realize you weren't using those components while the warnings are coming from the core filter/table. It's taking me some additional time to setup the MRI pipeline and so far I can only visit /electrophysiology_browser for testing.

**Review by maltheism — APPROVED** [review]
I'm approving the javascript part of the PR. This PR is large considering all of the work done. Other things will be necessary like handling errors but that can be done in separate PRs after this PR is merged.

**Comment by christinerogers** [comment]
@driusan - ready for your review -- this has 2 approvals and is needed for the release. thanks!

**Pull request merged**
